### ISSUE TITLE
Content with "</head>" as part of header leads to break?

### ISSUE THREAD

**Issue by spekulatius** [silverstripe/silverstripe-framework#9662]
Hey, 

a bit of a strange error/thing I've found here. More below.

## Affected Version

```
$ composer info --direct

innoweb/silverstripe-social-metadata 5.1.2  Adds social metadata like OpenGraph, Twitter and JSON-LD to all pages.
jonom/silverstripe-betternavigator   4.1.1  Front-end utility menu for SilverStripe websites featuring administration and development tools
phpunit/phpunit                      5.7.27 The PHP Unit Testing framework.
silverstripe/blog                    3.5.1  A fresh take on blogging in Silverstripe set out to tackle the issue of a cluttered Site Tree.
silverstripe/googlesitemaps          2.1.6  SilverStripe support for the Google Sitemaps XML, enabling Google and other search engines to see all urls on...
silverstripe/recipe-cms              4.6.1  SilverStripe recipe for fully featured page and asset content editing
silverstripe/recipe-plugin           1.3.0  Helper plugin to install SilverStripe recipes
symbiote/silverstripe-queuedjobs     4.6.0  A framework for defining and running background jobs in a queued manner
```

```
$ composer info | grep "silverstripe/"

silverstripe/admin                             1.6.1   SilverStripe admin interface
silverstripe/asset-admin                       1.6.1   Asset management for the SilverStripe CMS
silverstripe/assets                            1.6.1   SilverStripe Assets component
silverstripe/blog                              3.5.1   A fresh take on blogging in Silverstripe set out to tackle the issue of a cluttered Site Tree.
silverstripe/campaign-admin                    1.6.0   SilverStripe campaign admin interface
silverstripe/cms                               4.6.0   The SilverStripe Content Management System
silverstripe/config                            1.0.18  SilverStripe configuration based on YAML and class statics
silverstripe/errorpage                         1.6.0   ErrorPage component for SilverStripe CMS
silverstripe/framework                         4.6.2   The SilverStripe framework
silverstripe/googlesitemaps                    2.1.6   SilverStripe support for the Google Sitemaps XML, enabling Google and other search engines to se...
silverstripe/graphql                           3.3.0   GraphQL server for SilverStripe models and other data
silverstripe/lumberjack                        2.0.2   A module to make managing pages in a GridField easy without losing any of the functionality that...
silverstripe/mimevalidator                     2.0.0   Checks uploaded file content roughly matches a known MIME type for the file extension.
silverstripe/recipe-cms                        4.6.1   SilverStripe recipe for fully featured page and asset content editing
silverstripe/recipe-core                       4.6.1   SilverStripe framework-only core recipe
silverstripe/recipe-plugin                     1.3.0   Helper plugin to install SilverStripe recipes
silverstripe/reports                           4.6.0   Reports module for SilverStripe CMS
silverstripe/siteconfig                        4.6.0   Site wide settings administration.
silverstripe/tagfield                          2.4.3   Tag field for SilverStripe
silverstripe/vendor-plugin                     1.4.1   Allows vendor modules to expose directories to the webroot
silverstripe/versioned                         1.6.2   SilverStripe Versioned component
silverstripe/versioned-admin                   1.6.0   SilverStripe versioned admin interface
```

## Description

I'm not entirely sure, but I assume this is a framework issue. Google alerted me about broken structured data on one particular page:

![image](https://user-images.githubusercontent.com/8433587/91458111-3333e980-e896-11ea-9fa5-f824ece6a516.png)

The website is using framework v4.6.2., CMS v4.6.0 and [silverstripe-social-metadata](https://github.com/xini/silverstripe-social-metadata) v5.1.2 (latest) by @xini. It generates a JSON with meta information in the head tag (http://json-ld.org) to help with SEO.

On one page (which contains a code snippet with `</head>` in the content) the JSON breaks. Surprisingly it breaks right in the *middle* of the JSON - at the position of the closing head-tag. Maybe there is some trigger caused by the `</head>`? I'm a bit unsure about the reason.

## Steps to Reproduce

Please see the JSON in the head-tag. The attribute `articleBody` on [https://bringyourownideas.com/blog/passwordless-login-with-laravel/](https://bringyourownideas.com/blog/passwordless-login-with-laravel/) shows the issue.

Cheers!

**Comment by xini** [comment]
Thanks for tagging me @spekulatius. 
For the 'articleBody' I am using ->Plain() (https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/FieldType/DBHTMLText.php#L219) on the Content DB Object (https://github.com/xini/silverstripe-social-metadata/blob/master/src/Extensions/BlogPostExtension.php#L95), which should really give back a plain text version of the content, right?
The question is what a plain text version of a code example looks like? Not sure how I would solve this. But I think it should be solved in DBHTMLText->Plain()?
I am happy to implement a (temporary) workaround into my module if you happen to find one? ;)
Cheers

**Comment by spekulatius** [comment]
Hey @xini, 

thanks for replying! I assume it's caused by the closing head tag. Otherwise the JSON wouldn't break I guess. 

I'll download the live db and see what the Plain-output looks like later. Maybe this leads to a better understanding.

Cheers,
Peter

**Comment by xini** [comment]
Hey @spekulatius 
I think it's just a matter of correctly encoding the value. I'll set up a test site and have a look. 
Cheers

**Comment by spekulatius** [comment]
Yeah, that might be the issue. I might have trusted the WYSIWYG editor to much with this

**Comment by sminnee** [comment]
There's code that inserts JavaScript tags before the /head that is very simple - it's just a regex, for performance. This can be either disabled or I think replaced with a smarter one - check the Requirements API docs I think?

In 99% of cases it works so we allow for a config change in the 1%, which you found. ;-)

**Comment by spekulatius** [comment]
Hey @sminnee 

yeah, exactly this was the issue. I've played around and added a negative lookahead to the regex. With this it works for me. The linked PR would resolve this. I would appreciate if someone could test this to ensure it doesn't break anything I'm unaware off.

Thanks for your help Sam!

Cheers,
Peter

**Comment by sminnee** [comment]
Cool, the main thing about improving a regex is confirming that we have a decent test suite for it.

**Comment by sminnee** [comment]
On your site in question is there a `setWriteJavascriptToBody(false)` in PHP or `WriteJavascriptToBody: false` in YAML set on the Requirements_Backend in one of the modules? If so, that's the root cause of the issue here and you may want to investigate _why_. The insert-at-end-of-head code has some bugs, but it also leads to lower front-end performance.

**Comment by spekulatius** [comment]
Hey @sminnee,

I haven't found any YAML or PHP setting the parameter. While looking further I've noticed it's actually CSS files:

```
<link rel="stylesheet" type="text/css" href="/resources/vendor/silverstripe/blog/client/dist/styles/main.css?m=1591928642" />
<link rel="stylesheet" type="text/css" href="/resources/vendor/jonom/silverstripe-betternavigator/css/betternavigator.css?m=1523054468" />
</head>
```

The config only applies to JS files afaik.

I guess that's the reason why this isn't working and brings us back to square one, doesn't it?

Cheers,
Peter

**Comment by xini** [comment]
Hm. You're right. https://github.com/silverstripe/silverstripe-framework/blob/4/src/View/Requirements_Backend.php#L963 is the culprit.

**Comment by spekulatius** [comment]
Yeah, it is @xini. But getting all cases covered safely is a bit annoying - my PR is covering only one part of it.

**Comment by sminnee** [comment]
Oh right yes CSS file refs can't be moved to the body!

Fixing the escaping of the social module, if possible, or moving to a DOMDocument parser (optionally) seem like the ways forward

**Comment by spekulatius** [comment]
> Fixing the escaping of the social module, if possible, or moving to a DOMDocument parser (optionally) seem like the ways forward

Yeah, sounds like these are the only ways to fix it. My personal preference would be a DOM parser as it would fix the issue once and for all - but it's also more work :D

**Comment by xini** [comment]
I have fixed the social module and removed articleBody from the blog post json+ld to remove this duplication of content. articleBody is not required according to the Google guidelines: https://developers.google.com/search/docs/data-types/article.

**Comment by spekulatius** [comment]
Cheers @xini! Good to see we found a solution! 

I guess this very edge case is barely worth investing more time into it. Closing it for now.
